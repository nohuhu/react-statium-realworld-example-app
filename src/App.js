import React from 'react';
import { Routes, Route, useNavigate } from 'react-router-dom';
import Store from 'statium';

import getApi from './api.js';
import { setUser } from './actions/user.js';

import LoadMask from './components/LoadMask.js';
import Header from './components/Header.js';
import Footer from './components/Footer.js';
import Home from './components/Home/Home.js';
import Login from './components/Login.js';
import Logout from './components/Logout.js';
import Register from './components/Register.js';
import Article from './components/Article/ArticlePage.js';
import Editor from './components/Editor/Editor.js';
import Profile from './components/Profile.js';
import Settings from './components/Settings.js';

const initialState = {
  user: null,
  api: getApi(),
  appReady: false,
};

const App = () => {
  const navigate = useNavigate();

  return (
    <Store tag="App" data={{ navigate }} state={initialState} onMount={initialize}>
      {({ state }) => {
        if (!state.appReady) {
          return (
            <div className="container">
              <LoadMask loading={true} />
            </div>
          );
        }

        return (
          <>
            <Header />

            <Routes>
              <Route path="/login" element={<Login />} />
              <Route path="/logout" element={<Logout />} />
              <Route path="/register" element={<Register />} />
              <Route path="/article/:slug" element={<Article />} />
              <Route path="/editor/:slug" element={<Editor />} />
              <Route path="/editor" element={<Editor />} />
              <Route path="/settings" element={<Settings />} />
              <Route path="/@:username/favorites" element={<Profile tab="favorites" />} />
              <Route path="/@:username" element={<Profile tab="authored" />} />
              <Route path="/tag/:selectedTag" element={<Home tab="tag" />} />
              <Route path="/feed" element={<Home tab="feed" />} />
              <Route path="/" element={<Home tab="global" />} />
            </Routes>

            <Footer />
          </>
        );
      }}
    </Store>
  );
}

const initialize = async ({ set, dispatch }) => {
  const token = window.localStorage.getItem('jwtToken');

  if (token) {
    try {
      const api = getApi(token);
      const user = await api.User.current();

      await set({ api });
      await dispatch(setUser, user);
    }
    catch (e) {
      // Token expired, etc. Default API is tokenless, so no need to set it again.
      await dispatch(setUser, null);
    }
  }

  // User was not logged in previously, or token has expired.
  // Default state represents non-logged in experience so we are ready at this point.
  await set({ appReady: true });
};

export default App;
