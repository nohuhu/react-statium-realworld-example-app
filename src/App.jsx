import React from 'react';
import { Routes, Route, useNavigate } from 'react-router-dom';
import Store from 'statium';

import getApi from './api.js';
import { getUser, setUser } from './actions/user.js';

import LoadMask from './components/LoadMask.jsx';
import Header from './components/Header/Header.jsx';
import Footer from './components/Footer.jsx';
import Notifications from './components/NotificationProvider.jsx';
import { Alerts } from './components/Alerts.jsx'

import Home from './pages/Home.jsx';
import Login from './pages/Login.jsx';
import Logout from './pages/Logout.jsx';
import Register from './pages/Register.jsx';
import Article from './pages/Article.jsx';
import Editor from './pages/Editor.jsx';
import Profile from './pages/Profile.jsx';
import Settings from './pages/Settings.jsx';

const initialState = {
  user: null,
  api: getApi(),
  appReady: false,
};

const App = () => {
  // navigate() function is used in many actions, and the only way
  // to get the reference to this function is to call useNavigate
  // hook, which is only available at rendering time in functional
  // components, as per the rules of hooks.
  // One way to work with this limitation is keep track of every
  // action that needs to call navigate(), make sure that every
  // component dispatching such action is a functional component,
  // call useNavigate() in each of those components and pass the
  // reference to `navigate` function manually to every action
  // that makes use of it. Needless to say, this approach is
  // _really_ cumbersome.
  // So instead we are grabbing the reference to `navigate` once
  // and passing it to all components and actions down the tree
  // via the Statium Store `data` prop. It's as easy as that.
  const navigate = useNavigate();

  return (
    <Store tag="App" data={{ navigate }} state={initialState} onMount={initialize}>
    {({ state }) => {
      if (!state.appReady) {
        return (
          <div className="container">
            <LoadMask loading={true} />
          </div>
        );
      }

      return (
        <Notifications>
          <Header />
          <Alerts />

          <Routes>
            <Route path="/login" element={<Login />} />
            <Route path="/logout" element={<Logout />} />
            <Route path="/register" element={<Register />} />
            <Route path="/article/:slug" element={<Article />} />
            <Route path="/editor/:slug" element={<Editor />} />
            <Route path="/editor" element={<Editor />} />
            <Route path="/settings" element={<Settings />} />
            <Route path="/@:username/favorites" element={<Profile tab="favorites" />} />
            <Route path="/@:username" element={<Profile tab="authored" />} />
            <Route path="/tag/:selectedTag" element={<Home tab="tag" />} />
            <Route path="/feed" element={<Home tab="feed" />} />
            <Route path="/" element={<Home tab="global" />} />
          </Routes>

          <Footer />
        </Notifications>
      );
    }}
    </Store>
  );
}

const initialize = async ({ set, dispatch }) => {
  // Get the user record from window local storage token,
  // and validate it with the back end. If no token was stored
  // or is invalid, user will be null and we will use tokenless
  // API for non-authenticated experience.
  const user = await dispatch(getUser);

  await dispatch(setUser, user);

  await set({ appReady: true });
};

export default App;
