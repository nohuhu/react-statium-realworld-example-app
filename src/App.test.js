import React from 'react';
import { mount } from 'test/enzyme.js';

import App from './App';

describe("Application", () => {
  it("renders", () => {
    const tree = mount(<App />);

    expect(tree.find('App')).toMatchSnapshot();
  });
});
