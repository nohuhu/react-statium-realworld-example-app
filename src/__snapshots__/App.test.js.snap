// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application renders 1`] = `
<App>
  <Store
    data={Object {}}
    onMount={[Function]}
    state={
      Object {
        "api": Object {
          "Articles": Object {
            "all": [Function],
            "byAuthor": [Function],
            "bySlug": [Function],
            "byTag": [Function],
            "create": [Function],
            "delete": [Function],
            "deleteComment": [Function],
            "favorite": [Function],
            "favoritedBy": [Function],
            "feed": [Function],
            "get": [Function],
            "getComments": [Function],
            "postComment": [Function],
            "unfavorite": [Function],
            "update": [Function],
          },
          "Profile": Object {
            "follow": [Function],
            "get": [Function],
            "unfollow": [Function],
          },
          "Tags": Object {
            "all": [Function],
          },
          "User": Object {
            "current": [Function],
            "login": [Function],
            "register": [Function],
            "save": [Function],
          },
        },
        "appReady": false,
        "user": null,
      }
    }
    tag="App"
  >
    <div
      className="container"
    >
      <LoadMask
        loading={true}
      >
        <div
          className="load-mask "
        >
          <div
            className="load-mask-inner"
          >
            <i
              className="ion ion-load-c load-spinner"
              style={
                Object {
                  "fontSize": "50px",
                }
              }
            />
          </div>
        </div>
      </LoadMask>
    </div>
  </Store>
</App>
`;
